cdap-data-fabric/src/main/java/co/cask/cdap/data/format/AvroRecordFormat.java

```diff
@@ -16,36 +16,51 @@
 
 package co.cask.cdap.data.format;
 
+import co.cask.cdap.api.common.Bytes;
 import co.cask.cdap.api.data.format.UnexpectedFormatException;
 import co.cask.cdap.api.data.schema.Schema;
 import co.cask.cdap.api.data.schema.UnsupportedTypeException;
 import co.cask.cdap.api.flow.flowlet.StreamEvent;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.io.ByteBuffers;
 import co.cask.common.io.ByteBufferInputStream;
 import org.apache.avro.SchemaParseException;
 import org.apache.avro.generic.GenericDatumReader;
 import org.apache.avro.generic.GenericRecord;
 import org.apache.avro.io.BinaryDecoder;
-import org.apache.avro.io.DatumReader;
 import org.apache.avro.io.DecoderFactory;
 
 import java.io.IOException;
-import java.io.InputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
 import java.util.Map;
 
 /**
  * Stream record format that interprets the body as avro encoded binary data.
  */
 public class AvroRecordFormat extends StreamEventRecordFormat<GenericRecord> {
-  private org.apache.avro.Schema avroSchema;
-  private DecoderFactory decoderFactory = DecoderFactory.get();
-  private DatumReader<GenericRecord> datumReader;
+  private final ByteBufferInputStream byteBufferInput = new ByteBufferInputStream(ByteBuffers.EMPTY_BUFFER);
+  private final DecoderFactory decoderFactory = DecoderFactory.get();
+
+  private org.apache.avro.Schema readSchema;
+  private String writeSchemaHash;
+  private GenericDatumReader<GenericRecord> datumReader;
   private BinaryDecoder binaryDecoder;
 
   @Override
   public GenericRecord read(StreamEvent event) {
     try {
-      return datumReader.read(null, decoderFactory.binaryDecoder(new ByteBufferInputStream(event.getBody()),
-                                                                 binaryDecoder));
+      // Check if the event has different schema then the read schema. If it does update the datumReader
+      String eventSchemaStr = event.getHeaders().get(Constants.Stream.Headers.SCHEMA);
+      String eventSchemaHash = event.getHeaders().get(Constants.Stream.Headers.SCHEMA_HASH);
+      if (eventSchemaStr != null && !this.writeSchemaHash.equals(eventSchemaHash)) {
+        org.apache.avro.Schema writeSchema = new org.apache.avro.Schema.Parser().parse(eventSchemaStr);
+        datumReader.setSchema(writeSchema);
```

Comment:
Alternatively we can switch back to use the read schema as the default one if the event doesn't have schema, which probably match better with existing logic.